### DISCUSSION TITLE
Simulation Study on Ensemble Classifiers

### DISCUSSION THREAD

**Discussion by rohitgarud** [asreview/asreview#1334] [Show and tell]
Results of a simulation study are presented in the following table, where an ensemble classifier using Naive Bayes (NB) and Logistic Regression was developed and its performance on benchmark datasets as compared to the NB classifier is studied for the TFIDF features.  The default max query strategy was used. Two ensembling strategies are proposed. 
- Multiply Ensemble Strategy:
In this strategy, the probabilities at each iteration are calculated using NB and Logistic classifiers, and the probabilities are then multiplied to get the final value. This allows for the records for which both the models have an agreement and high confidence to rank higher. If one classifier gives a probability close to 1 (relevant) and the other gives a probability close to 0 (irrelevant), this gives a lower value than for the case where both the classifiers have assigned relevant probability. 
- Random Ensemble Strategy
In this strategy, a classifier is randomly selected at each iteration and the probabilities are calculated using the selected classifier. As the sequence of records found by different classifiers is different, this allows for possibly faster discovery, and the ensembled models can work complementing each other (?).

In many cases, the ensemble models had a higher recall and WSS while having lower ATD.

The ensemble models are developed as ASReview model extension and can be downloaded/installed from the [Asreview-Ensemble-Classifiers Repository](https://github.com/rohitgarud/Asreview-Ensemble-Classifiers). Simulations were performed using the awesome [ASReview Makita Extension](https://github.com/asreview/asreview-makita) with 'multiple model template'.
After installing the ensemble model extension and asreview-makita extension, we can run following command to get the jobs file (for windows).
```
asreview makita template multiple_models --classifiers nb ensemble_nb_lr --feature_extractors tfidf -f jobs.bat
```
and then run `jobs.bat` command to run the simulations.

**Comment by rohitgarud** [comment]
| Dataset                   | Classifier             | Ensemble Strategy | recall_0.1 | recall_0.5 | recall_0.9 | wss_0.95 | erf_0.1 | atd      |
|---------------------------|------------------------|-------------------|------------|------------|------------|----------|---------|----------|
| ACE Inhibitors            | Ensemble NB + Logistic | Multiply          | 0.85       | 0.975      | 0.975      | 0.789    | 0.75    | 177.75   |
|                           |                        | Random            | 0.85       | 0.975      | 0.975      | 0.778    | 0.75    | 183.975  |
|                           | NB                     |                   | 0.85       | 0.95       | 1          | 0.782    | 0.75    | 190.325  |
| ADHD                      | Ensemble NB + Logistic | Multiply          | 0.789      | 0.895      | 1          | 0.411    | 0.684   | 97.053   |
|                           |                        | Random            | 0.737      | 0.947      | 1          | 0.456    | 0.632   | 93.789   |
|                           | NB                     |                   | 0.842      | 0.947      | 1          | 0.446    | 0.737   | 96.789   |
| Antihistamines            | Ensemble NB + Logistic | Multiply          | 0.333      | 0.867      | 0.933      | 0.055    | 0.267   | 73.133   |
|                           |                        | Random            | 0.267      | 0.867      | 0.933      | 0.094    | 0.2     | 77.667   |
|                           | NB                     |                   | 0.333      | 0.867      | 0.933      | 0.091    | 0.267   | 72.333   |
| Appenzeller-Herzog_2020   | Ensemble NB + Logistic | Multiply          | 0.786      | 1          | 1          | 0.779    | 0.679   | 216.857  |
|                           |                        | Random            | 0.786      | 1          | 1          | 0.75     | 0.679   | 236.071  |
|                           | NB                     |                   | 0.821      | 0.964      | 1          | 0.723    | 0.714   | 263.464  |
| Atypical Antipsychotics   | Ensemble NB + Logistic | Multiply          | 0.386      | 0.876      | 0.979      | 0.231    | 0.29    | 266.69   |
|                           |                        | Random            | 0.352      | 0.876      | 0.972      | 0.242    | 0.255   | 268.4    |
|                           | NB                     |                   | 0.352      | 0.869      | 0.979      | 0.168    | 0.255   | 275.124  |
| Bannach-Brown_2019        | Ensemble NB + Logistic | Multiply          | 0.545      | 0.95       | 0.996      | 0.468    | 0.444   | 277.226  |
|                           |                        | Random            | 0.527      | 0.953      | 0.993      | 0.482    | 0.427   | 271.111  |
|                           | NB                     |                   | 0.523      | 0.932      | 0.989      | 0.404    | 0.423   | 293.022  |
| Beta Blockers             | Ensemble NB + Logistic | Multiply          | 0.659      | 0.951      | 1          | 0.655    | 0.561   | 234.317  |
|                           |                        | Random            | 0.683      | 0.927      | 1          | 0.641    | 0.585   | 216.951  |
|                           | NB                     |                   | 0.707      | 0.927      | 0.976      | 0.669    | 0.61    | 241.878  |
| Bos_2018                  | Ensemble NB + Logistic | Multiply          | 1          | 1          | 1          | 0.83     | 0.9     | 89.4     |
|                           |                        | Random            | 1          | 1          | 1          | 0.832    | 0.9     | 91.8     |
|                           | NB                     |                   | 1          | 1          | 1          | 0.834    | 0.9     | 76       |
| Calcium Channel Blockers  | Ensemble NB + Logistic | Multiply          | 0.303      | 0.919      | 0.99       | 0.245    | 0.202   | 244.232  |
|                           |                        | Random            | 0.263      | 0.919      | 0.98       | 0.32     | 0.162   | 259.697  |
|                           | NB                     |                   | 0.273      | 0.899      | 0.96       | 0.229    | 0.172   | 268.364  |
| Estrogens                 | Ensemble NB + Logistic | Multiply          | 0.278      | 0.899      | 0.975      | 0.257    | 0.177   | 89.152   |
|                           |                        | Random            | 0.291      | 0.899      | 0.962      | 0.339    | 0.19    | 89.62    |
|                           | NB                     |                   | 0.278      | 0.911      | 0.962      | 0.249    | 0.177   | 89.418   |
| Hall_2012                 | Ensemble NB + Logistic | Multiply          | 0.99       | 1          | 1          | 0.912    | 0.893   | 133.272  |
|                           |                        | Random            | 0.99       | 1          | 1          | 0.911    | 0.893   | 139.058  |
|                           | NB                     |                   | 0.99       | 1          | 1          | 0.911    | 0.893   | 134.272  |
| Kitchenham_2010           | Ensemble NB + Logistic | Multiply          | 0.318      | 0.977      | 1          | 0.639    | 0.227   | 255.795  |
|                           |                        | Random            | 0.568      | 0.977      | 1          | 0.604    | 0.477   | 240.75   |
|                           | NB                     |                   | 0.659      | 0.977      | 1          | 0.652    | 0.568   | 200.136  |
| Kwok_2020                 | Ensemble NB + Logistic | Multiply          | 0.647      | 0.983      | 1          | 0.672    | 0.546   | 225.605  |
|                           |                        | Random            | 0.664      | 0.983      | 1          | 0.671    | 0.563   | 235.059  |
|                           | NB                     |                   | 0.689      | 0.992      | 1          | 0.675    | 0.588   | 227.092  |
| Nagtegaal_2019            | Ensemble NB + Logistic | Multiply          | 0.65       | 1          | 1          | 0.532    | 0.55    | 219.83   |
|                           |                        | Random            | 0.64       | 0.98       | 1          | 0.48     | 0.54    | 221.09   |
|                           | NB                     |                   | 0.64       | 0.99       | 1          | 0.529    | 0.54    | 237.19   |
| NSAIDS                    | Ensemble NB + Logistic | Multiply          | 0.575      | 0.975      | 1          | 0.696    | 0.475   | 44.95    |
|                           |                        | Random            | 0.525      | 0.975      | 1          | 0.691    | 0.425   | 48.125   |
|                           | NB                     |                   | 0.6        | 0.975      | 0.975      | 0.698    | 0.5     | 46.775   |
| Opiods                    | Ensemble NB + Logistic | Multiply          | 0          | 0.929      | 1          | 0.482    | -0.071  | 757.214  |
|                           |                        | Random            | 0          | 0.929      | 1          | 0.444    | -0.071  | 817.071  |
|                           | NB                     |                   | 0          | 0.929      | 1          | 0.467    | -0.071  | 717.857  |
| Oral Hypoglycemics        | Ensemble NB + Logistic | Multiply          | 0.259      | 0.778      | 0.985      | 0.218    | 0.163   | 150.437  |
|                           |                        | Random            | 0.274      | 0.778      | 0.985      | 0.16     | 0.178   | 152.304  |
|                           | NB                     |                   | 0.237      | 0.77       | 0.978      | 0.19     | 0.141   | 155.444  |
| Proton Pump Inhibitors    | Ensemble NB + Logistic | Multiply          | 0.72       | 0.94       | 1          | 0.47     | 0.62    | 196.52   |
|                           |                        | Random            | 0.62       | 0.94       | 1          | 0.507    | 0.52    | 191.52   |
|                           | NB                     |                   | 0.66       | 0.94       | 1          | 0.481    | 0.56    | 191.96   |
| Radjenovic_2013           | Ensemble NB + Logistic | Multiply          | 0.957      | 1          | 1          | 0.866    | 0.851   | 175.851  |
|                           |                        | Random            | 0          | 0.021      | 0.128      | -0.074   | -0.106  | 5724.213 |
|                           | NB                     |                   | 0.957      | 1          | 1          | 0.841    | 0.851   | 189.574  |
| Skeletal Muscle Relaxants | Ensemble NB + Logistic | Multiply          | 0          | 0.625      | 0.75       | -0.091   | -0.125  | 752.875  |
|                           |                        | Random            | 0          | 0.625      | 0.75       | -0.091   | -0.125  | 795.125  |
|                           | NB                     |                   | 0          | 0.625      | 1          | -0.015   | -0.125  | 782.625  |
| Statins                   | Ensemble NB + Logistic | Multiply          | 0.738      | 0.94       | 0.988      | 0.527    | 0.643   | 443.024  |
|                           |                        | Random            | 0.726      | 0.94       | 0.988      | 0.589    | 0.631   | 422.25   |
|                           | NB                     |                   | 0.726      | 0.94       | 1          | 0.471    | 0.631   | 445.036  |
| Triptans                  | Ensemble NB + Logistic | Multiply          | 0.565      | 0.957      | 1          | 0.498    | 0.478   | 95.652   |
|                           |                        | Random            | 0.565      | 0.957      | 1          | 0.45     | 0.478   | 104.478  |
|                           | NB                     |                   | 0.565      | 0.957      | 1          | 0.52     | 0.478   | 98.565   |
| Urinary Incontinence      | Ensemble NB + Logistic | Multiply          | 0.41       | 0.949      | 1          | 0.455    | 0.308   | 60.231   |
|                           |                        | Random            | 0.487      | 0.949      | 1          | 0.471    | 0.385   | 54.744   |
|                           | NB                     |                   | 0.462      | 0.872      | 1          | 0.382    | 0.359   | 60.923   |
| van_de_Schoot_2017        | Ensemble NB + Logistic | Multiply          | 1          | 1          | 1          | 0.889    | 0.905   | 111.31   |
|                           |                        | Random            | 1          | 1          | 1          | 0.882    | 0.905   | 118.762  |
|                           | NB                     |                   | 1          | 1          | 1          | 0.888    | 0.905   | 101.571  |
| van_Dis_2020              | Ensemble NB + Logistic | Multiply          | 0.625      | 0.986      | 1          | 0.63     | 0.528   | 1144.153 |
|                           |                        | Random            | 0.653      | 0.986      | 1          | 0.63     | 0.556   | 1189.486 |
|                           | NB                     |                   | 0.653      | 0.972      | 1          | 0.65     | 0.556   | 1191.264 |
| Wahono_2015               | Ensemble NB + Logistic | Multiply          | 0.967      | 1          | 1          | 0.863    | 0.869   | 211.951  |
|                           |                        | Random            | 0.967      | 1          | 1          | 0.86     | 0.869   | 222.115  |
|                           | NB                     |                   | 0.967      | 1          | 1          | 0.863    | 0.869   | 197.803  |
| Wolters_2018              | Ensemble NB + Logistic | Multiply          | 0.889      | 1          | 1          | 0.807    | 0.778   | 172.889  |
|                           |                        | Random            | 0.889      | 1          | 1          | 0.781    | 0.778   | 179.056  |
|                           | NB                     |                   | 0.889      | 1          | 1          | 0.794    | 0.778   | 162.222  |

**Reply by Rensvandeschoot** [reply]
This is an exciting model and a great simulation study!! Will you publish these results, for example, on a pre-print server so that you receive a digital object identifier (DOI) and we can cite your work?

**Reply by rohitgarud** [reply]
I was not planning to publish a paper myself as I was just fidling with the simulation mode and asreview-makita, but if anybody is interested in co-authoring (the writing part) then I am happy to assist because I currently, unfortunately, can not spend time on the paper writing.

**Reply by J535D165** [reply]
Very nice work!! Impressive modeling.

**Reply by rohitgarud** [reply]
Hello @Rensvandeschoot and @J535D165. I am planning to write a paper and publish it in Arxiv as you suggested. Can you please suggest which benchmark datasets I should include in the results because, for a few of them,  all the models, including default Naive Bayes, perform poorly. 
Thank you.

**Reply by Rensvandeschoot** [reply]
good question... You could either use all of the datasets and state that the performance of NB is not optimal for some datasets and more research is needed to find out why (there can be many reasons related to dataset characteristics or technical reasons like noisy labels, missing data, etc - for this reason we are planning to upgrade the benchmark dataset repository soon). Alternatively, you can create a sub-set, but the subset selection should not be based on the performance of the models to avoid selection bias.

**Reply by rohitgarud** [reply]
Thank you @Rensvandeschoot for your suggestions. I think I will wait for the upgrade and perform the simulations again, so that the simulation study will be inline with the benchmark datasets and possible future developments in the same direction will have a good starting point to refer to.

**Reply by Rensvandeschoot** [reply]
yes, this sounds as a good idea!
